File: FinalProject/server/book_smart/models/book.py
"""
book.py

Created by Stephen Andrews, February 17, 2018.
"""

from book_smart.extensions import db

author_identifier = db.Table('author_identifier',
    db.Column('book_isbn', db.Text, db.ForeignKey('book.isbn')),
    db.Column('author_id', db.Integer, db.ForeignKey('author.author_id'))
)

class Book(db.Model):
    """Represents a book."""

    isbn = db.Column(db.Text, primary_key=True, unique=True, nullable=False)
    authors = db.relationship('Author', secondary=author_identifier)
    description = db.Column(db.Text, nullable=False)
    publisher = db.Column(db.Text, nullable=False)
    published_date = db.Column(db.Integer, nullable=False)
    title = db.Column(db.Text, nullable=False)
    thumbnail = db.Column(db.Text, nullable=False)
    small_thumbnail = db.Column(db.Text, nullable=False)
    subtitle = db.Column(db.Text)

    def to_json(self):
        book = {}
        book['isbn'] = self.isbn
        book['authors'] = ['{first_name} {last_name}'.format(**author.to_json()) for author in self.authors]
        book['description'] = self.description
        book['publisher'] = self.publisher
        book['publishedDate'] = self.published_date
        book['title'] = self.title
        book['thumbnail'] = self.thumbnail
        book['smallThumbnail'] = self.small_thumbnail
        book['subtitle'] = self.subtitle

        return book

File: FinalProject/WPIReads/app/src/main/java/cs4518_team6/booksmart/model/Book.java
package cs4518_team6.booksmart.model;

import android.util.Log;

import com.github.kevinsawicki.http.HttpRequest;
import com.google.gson.Gson;

import java.net.HttpURLConnection;
import java.util.Arrays;

import cs4518_team6.booksmart.Constants;

/**
 * Represents a book.
 * Created by Stephen Andrews.
 */
public class Book extends Model {

    private static final String ENDPOINT = "/books/";
    private String isbn;
    private String[] authors;
    private String description;
    private String publisher;
    private String publishedDate;
    private String title;
    private String thumbnail;
    private String smallThumbnail;
    private String subtitle;

    public Book(String isbn, String[] authors, String description, String publisher,
                String publishedDate, String title, String thumbnail, String smallThumbnail, String subtitle) {
        this.isbn = isbn;
        this.authors = authors;
        this.description = description;
        this.publisher = publisher;
        this.publishedDate = publishedDate;
        this.title = title;
        this.thumbnail = thumbnail;
        this.smallThumbnail = smallThumbnail;
        this.subtitle = subtitle;
    }

    /**
     * Get a book for a specified ISBN-10 identifier.
     * @param isbn The isbn of the book to get.
     * @return The {@link Book} or null if invalid ISBN.
     */
    public static Book get(String isbn) {
        String url = Constants.API_URL + ENDPOINT + isbn;
        HttpRequest request = HttpRequest.get(url);
        String data = request.body();
        Log.d("steve", data);

        switch (request.code()) {
            case HttpURLConnection.HTTP_OK:
                Gson gson = new Gson();
                return gson.fromJson(data, Book.class);
            default:
                return null;
        }
    }

    public String getIsbn() {
        return isbn;
    }

    public String[] getAuthors() {
        return authors;
    }

    public String getDescription() {
        return description;
    }

    public String getPublisher() {
        return publisher;
    }

    public String getPublishedDate() {
        return publishedDate;
    }

    public String getTitle() {
        return title;
    }

    public String getThumbnail() {
        return thumbnail;
    }

    public String getSmallThumbnail() {
        return smallThumbnail;
    }

    public String getSubtitle() {
        return subtitle;
    }

    @Override
    public String toString() {
        return "Book{" +
                "isbn='" + isbn + '\'' +
                ", authors=" + Arrays.toString(authors) +
                ", description='" + description + '\'' +
                ", publisher='" + publisher + '\'' +
                ", publishedDate='" + publishedDate + '\'' +
                ", title='" + title + '\'' +
                ", thumbnail='" + thumbnail + '\'' +
                ", smallThumbnail='" + smallThumbnail + '\'' +
                ", subtitle='" + subtitle + '\'' +
                '}';
    }
}
